[CN-1152]: add localConfig to persistence (#171)
hazelcast/hazelcast-platform-operator-docs: docs/modules/ROOT/pages/backup-restore.adoc

This topic assumes that you know about Persistence in Hazelcast. To learn about Persistence, see the xref:hazelcast:storage:persistence.adoc[Platform documentation].

> There are two options for a backup: local and external. Local backups are kept in volume and never moved anywhere whereas external backups are moved into buckets provided by the user.

Reviewer: ```suggestion
There are two options for a backup: local and external. Local backups are kept in volume and never moved. External backups are moved into buckets provided by the user.
```

Reviewer: or a list, something like:

Backups can be either of the following:

- Local. Local backups are....
- External. External backups are...

Author: https://github.com/hazelcast/hazelcast-platform-operator-docs/pull/171/commits/29e7229d15fce92bb46b030a72d16010c16fecd7

<1> `HotBackup` resource name used for restore. The backup folder name will be taken from the `HotBackup` resource.
<2> Agent responsible for restoring data from the local storage.
The agent configuration is optional. If you give `restore` under `persistence` and do not pass the agent configuration, Hazelcast Platform Operator
uses the latest agent version that is compatible with its version.

> NOTE: For restore to be successful you have to give restore configuration at the `Hazelcast` custom resource creation. Otherwise, triggering restore on currently running members will likely to fail.

Reviewer: ```suggestion
NOTE: For restore to be successful you must configure restore when creating the `Hazelcast` custom resource. Otherwise, triggering restore on currently running members is likely to fail.
```

> To restore from local backups that is not taken by HotBackup resource, configure `localConfig`.

Reviewer: ```suggestion
To restore from local backups that were not created by the HotBackup resource, configure `localConfig`.
```

[source,yaml,subs="attributes+"]
----
include::ROOT:example$/hazelcast-persistence-restore-local.yaml[]
----

> <1> For restore to be successful, name of the cluster that the backup taken and restored must be the same.

Reviewer: ```suggestion
<1> The name of the cluster for both backup and restore must be the same. If the cluster name is different, the restore fails.
```

<2> `baseDir` is the root directory of persistence.
<3> `backupDir` is the directory for backups.

> <4> `backupFolder` is the folder for a specific backup.

Reviewer: A question: what is the difference here between a directory and a folder? I've always considered them to be interchangeable terms for the structures that contain other structures and files, am I wrong?

Author: https://github.com/hazelcast/hazelcast-platform-operator-docs/pull/171/commits/cfaf24f84ded67a21f554fdbc79e7180a97d4c35

> NOTE: Agent copies backup to be restored from `baseDir` + `backupDir` + `backupFolder` to `/data/hot-restart` path.

Reviewer: ```suggestion
NOTE: The agent copies the backup to be restored from `baseDir`, `backupDir`, and `backupFolder` to the `/data/hot-restart` path.
```

Author: my intention while using `+` here is to explain that the agent joins those parameters, `,` does not give that idea IMO. Do you have any other idea?

Author: https://github.com/hazelcast/hazelcast-platform-operator-docs/pull/171/commits/42d63b960040c5395c3c8e7bf8d026dca387231a

<1> `HotBackup` resource name used for restore. The backup folder name will be taken from the `HotBackup` resource.
<2> Agent responsible for restoring data from the local storage.
The agent configuration is optional. If you give `restore` under `persistence` and do not pass the agent configuration, Hazelcast Platform Operator
uses the latest agent version that is compatible with its version.

> To restore from local backups that were not created by the HotBackup resource, configure `localConfig`.

Reviewer 2: We should have titles or at least listing approach to emphasize the two different restore scenario when we have local backup. I don't know if it makes sense for you but I suggest something like this.
- Local Backups taken with the Operator.
- Local Backups taken with the Helm Chart 

While thinking about their difference, I bring another idea. How about creating new page just for the newly added restore support? Since backup and restore features are expected as one of the main features which will be used always by the all customers who are using operator. However, newly supported restore is not actually a feature which will be used continuously. Instead, it's kinda migration solution for older operator versions and helm chart users. Also in the future we may decide to support fully custom backups. Are we going to put it also in this page? I don't think so, it'll be a migration solution as well so I suggest to separate new restore option from the existing one. Sorry for late feedback, I just got into this idea.

Author: Such a nice a feedback Cagri, let's talk with the team and decide together. I am not sure %100 but it might be a good idea to make it a part of migration documents.

Author: A better idea would be to put the documentation about localConfig in here, and reference it from migration documentations.

Author: I committed more changes to improve the docs: https://github.com/hazelcast/hazelcast-platform-operator-docs/pull/171/commits/32bb2ddc3815561064cb9d4cdf54d7fd49800fe4. I didn't mention about the helm and migration words explicitly, because in the feature we will have specific guides for those scenarios like https://github.com/hazelcast/hz-docs/pull/1052. We will refer to this document from them.

[source,yaml,subs="attributes+"]
----
include::ROOT:example$/hazelcast-persistence-restore-local.yaml[]
----

> <1> The name of the cluster for both backup and restore must be the same. If the cluster name is different, the restore fails.

Reviewer 2: Since there is a field called `ClusterName` in HZ CR. This sentence may confuse some people. I would go with `Hazelcase CR Name`. WDYT?

> <2> If you configured maps with persistence, you need to create them while creating Hazelcast CR  as well. To be able to create them while creating Hazelcast CR, you need to use xref:custom-config.adoc[Custom Config]. If you try to apply Hazelcast CR and then Map CRs, it will not work.

Reviewer: ```suggestion
<2> If you configured maps with persistence, you must create them while creating the Hazelcast CR. To create them while creating the Hazelcast CR, you must use xref:custom-config.adoc[Custom Config]. You cannot apply the Hazelcast CR and then the Map CRs.
```

Reviewer 2: I don't get the need for Custom Config for the persisted maps. What do we want to achieve?

> <3> `pvcNamePrefix` is the prefix of the existing PVCs. It can be `persistence` or `hot-restart-persistence` depending on the installation method (in different versions of Hazelcast Helm Chart and Hazelcast Platform Operators these two values are used by default) of the old cluster. You can decide this value by running `kubect get pvc` command.

Reviewer: ```suggestion
<3> `pvcNamePrefix` is the prefix of the existing PVCs. It can be set to `persistence` or `hot-restart-persistence` depending on the installation method of the old cluster. In different versions of Hazelcast Helm Chart and Hazelcast Platform Operator, these values are used by default. Run the `kubect get pvc` command to decide which setting best suits your requirements.
```

<4> `baseDir` is the root directory for persistence.
<5> `backupDir` is the directory that contains a backupFolder for each available backup.
<6> `backupFolder` is the directory containing the specific backup for the restore.

> TIP: To find the `backupFolder` value, you can run a simple pod that lists the contents of the PVC and check its logs, it will print the folder structure of the specified PVC. For example:

Reviewer: ```suggestion
TIP: To find the `backupFolder` value, you can run a simple pod that lists the contents of the PVC and checks its logs. This lists the folder structure of the specified PVC. For example:
```

[source,yaml,subs="attributes+"]
----
include::ROOT:example$/pod-local-pvc-content.yaml[]
----

> <1>  Replace `/data/persistence` with the path where your PV is mounted inside the container.

Reviewer: ```suggestion
<1>  Replace `/data/persistence` with the path to your PV, which is mounted inside the container.
```

<2> Replace `/data/persistence` with the correct mountPath specified in PV.

> <3> Replace with the name of the one of the PVCs mounted to the cluster that the backup is taken.

Reviewer: ```suggestion
<3> Replace with the name of the one of the PVCs, which is mounted to the cluster from which the backup is taken.
```

> NOTE: Agent copies backup to be restored from `{baseDir}/{backupDir}/{backupFolder}` to `/data/persistence/base-dir` path.

Reviewer: ```suggestion
NOTE: Agent copies the backup to be restored from `{baseDir}/{backupDir}/{backupFolder}` to `/data/persistence/base-dir`.
```

<1> `HotBackup` resource name used for the restore. The bucket URI and secret are taken from the `HotBackup` resource.
<2> Agent responsible for restoring data from external storage.
The agent configuration is optional. If you provide `restore` under `persistence` and do not pass the agent configuration, Hazelcast Platform Operator
uses the latest agent version that is compatible with its version.
--

====

> NOTE: For restore to be successful you have to give restore configuration at the `Hazelcast` custom resource creation. Otherwise, triggering restore on currently running members will likely to fail.

Reviewer: ```suggestion
NOTE: For restore to be successful you must configure restore when creating the `Hazelcast` custom resource. Otherwise, triggering restore on currently running members is likely to fail.
```
